src/core/mesh/qgsmeshdataprovider.h

```diff
@@ -117,6 +142,70 @@ class CORE_EXPORT QgsMeshDatasetValue
     double mY  = std::numeric_limits<double>::quiet_NaN();
 };
 
+/**
+ * \ingroup core
+ *
+ * QgsMeshDataBlock is a block that can be used to retreive a block of
+ * active flag (e.g. face's active integer flag)
+ * scalars (e.g. scalar dataset double values)
+ * vectors (e.g. vector dataset doubles x,y values)
+ *
+ * data are implicitely shared, so the class can be quickly copied
+ * std::numeric_limits<double>::quiet_NaN() represents NODATA value
+ *
+ * \since QGIS 3.6
+ */
+class CORE_EXPORT QgsMeshDataBlock
+{
+  public:
+    //! Type of data stored in the block
+    enum DataType
+    {
+      ActiveFlagInteger, //!< Integer boolean flag whether face is active
+      ScalarDouble, //!< Scalar double values
+      Vector2DDouble, //!< Vector double pairs (x1, y1, x2, y2, ... )
+    };
+
+    //! Constructs an invalid block
+    QgsMeshDataBlock();
+
+    //! Constructs a new block
+    QgsMeshDataBlock( DataType type, int count );
+
+    //! Type of data stored in the block
+    DataType type() const;
+
+    //! Number of items stored in the block
+    int count() const;
+
+    //! Whether the block is valid
+    bool isValid() const;
+
+    /**
+     * Returns a value represented by the index
+     * For active flag the behaviour is undefined
+     */
+    QgsMeshDatasetValue value( int index ) const;
+
+    /**
+     * Returns a value for active flag by the index
+     * For scalar and vector 2d the behaviour is undefined
+     */
+    bool active( int index ) const;
+
+    /**
+     * Returns internal buffer to the array
+     * The buffer is already allocated
+     * Use to retreive data from MDAL provider
+     */
+    void *buffer();
```

Comment:
Shall we also add `const void *constBuffer() const;` for efficient reading of data out of the block? (it would use constData() call on QVector instead of data() to avoid detach)